spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/pubsub/PubSubBeanProcessor.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2021-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spring.autoconfigure.pubsub;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import com.google.cloud.spring.autoconfigure.core.GcpProperties;
+import com.google.cloud.spring.core.DefaultGcpProjectIdProvider;
+import com.google.cloud.spring.core.GcpProjectIdProvider;
+import com.google.cloud.spring.pubsub.core.PubSubConfiguration;
+import com.google.cloud.spring.pubsub.support.PubSubSubscriptionUtils;
+
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.config.ConfigurableListableBeanFactory;
+import org.springframework.beans.factory.support.BeanDefinitionBuilder;
+import org.springframework.beans.factory.support.BeanDefinitionRegistry;
+import org.springframework.beans.factory.support.BeanDefinitionRegistryPostProcessor;
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.boot.context.properties.bind.Binder;
+import org.springframework.boot.context.properties.source.ConfigurationPropertySources;
+import org.springframework.core.env.ConfigurableEnvironment;
+import org.springframework.scheduling.concurrent.ThreadPoolTaskScheduler;
+
+/**
+ * Class to register ThreadPoolTaskScheduler beans.
+ */
+public class PubSubBeanProcessor implements BeanDefinitionRegistryPostProcessor {
+
+	private ConfigurableEnvironment environment;
+
+	private ConcurrentHashMap<String, ThreadPoolTaskScheduler> threadPoolTaskSchedulerMap = new ConcurrentHashMap<>();
+
+	private ThreadPoolTaskScheduler globalScheduler;
+
+	public PubSubBeanProcessor(ConfigurableEnvironment environment) {
+		this.environment = environment;
+	}
+
+	@Override
+	public void postProcessBeanDefinitionRegistry(BeanDefinitionRegistry beanDefinitionRegistry) throws BeansException {
+		Binder binder = new Binder(ConfigurationPropertySources.get(this.environment));
+		String cloudPropertiesPrefix = GcpProperties.class.getAnnotation(ConfigurationProperties.class)
+				.value();
+		GcpProperties gcpProperties = binder.bind(cloudPropertiesPrefix, GcpProperties.class)
```

Comment:
The unit test failure seems to be happening due to this because the GcpProjectIdProvider bean isn't accessible to this method.

Reply:
Fixed test my modifying test setup to specify projectId through application.properties. However, this seems to be a drawback of using `Binder` to determine the projectId in `BeanDefinitionRegistryPostProcessor`